### PR TITLE
Update 'test/setup' to work with Dispatch integrations

### PR DESCRIPTION
**Pull request by kierachell**
connects [QA #29](https://github.com/department-of-veterans-affairs/appeals-qa/issues/29)

Basically a big rework of `test/setup` and other "data reset code" to decouple it from production/application code. This will also enable me to schedule automated Dispatch smoke tests with minimal manual configuration.

The main goal is so that we can just delete all of the test code and the app will run fine.

- Updated form for Certifications reset code
- Dispatch data reset
  - Clear tasks/claims/appeals
  - Reset special issues, location, date and cancelling existing EPs for dispatched appeals (with **new** `TIOCTIME`!)
  - Link for creating new tasks
  - Full and Partial Grant handling!
- Tests to test our reset test data code! I did test it manually as well to make sure it "tries to call VACOLS/VBMS"
- Minor tweaks here and there because I couldn't help myself

I basically rewrote the app backwards...

### CODE REVIEW

File: app/models/appeal.rb [hunk 1]
```diff
@@ -153,11 +153,6 @@ def certify!
     Appeal.certify(self)
   end
 
-  def uncertify!(user_id)
```
Comment (shanear) [inline app/models/appeal.rb]: 👍 great that you are removing this 🔨

File: app/controllers/test/users_controller.rb [hunk 2]
```diff
@@ -1,6 +1,38 @@
...
+  # Set current user in DEMO
+  def set_user
+    User.before_set_user # for testing only
+
+    session["user"] = User.authentication_service.get_user_session(params[:id])
+    redirect_to "/test/users"
+  end
+
+  # Set end products in DEMO
+  def set_end_products
```
Comment (shanear) [inline app/controllers/test/users_controller.rb]: why did you move this here? Seems weird to be in `UsersController`

File: app/controllers/test/setup_controller.rb [hunk 3]
```diff
@@ -1,84 +1,77 @@
...
-    redirect_to "/test/users"
+    redirect_to action: "index"
   end
-  # :nocov:
 
   private
 
-  def require_test_user
+  # :nocov:
+  def require_uat
```
Comment (shanear) [inline app/controllers/test/setup_controller.rb]: why rename this function. It seems like it's still requiring the test user?

File: app/controllers/test/setup_controller.rb [hunk 4]
```diff
@@ -1,84 +1,77 @@
...
+  def certification_ids
+    ENV["TEST_APPEAL_IDS"].split(",")
   end
 
   def full_grant_ids
     ENV["FULL_GRANT_IDS"].split(",")
   end
 
-  def partial_grant_ids
+  def part_remand_ids
```
Comment (shanear) [inline app/controllers/test/setup_controller.rb]: what does `part_remand_ids` mean?

File: app/controllers/test/setup_controller.rb [hunk 5]
```diff
@@ -1,84 +1,77 @@
...
-    case params[:type]
-    when "full"
-      BGSService.end_product_data = BGSService.existing_full_grants
-    when "partial"
-      BGSService.end_product_data = BGSService.existing_partial_grants
-    when "none"
-      BGSService.end_product_data = BGSService.no_grants
-    when "all"
-      BGSService.end_product_data = BGSService.all_grants
+  def delete_test_data
```
Comment (shanear) [inline app/controllers/test/setup_controller.rb]: This one is pretty dangerous. Any way we can scope this or add some additional safeguards?

File: app/services/test_data_service.rb [hunk 6]
```diff
@@ -64,6 +65,36 @@ def self.cancel_end_products(appeal)
...
+          SET BFDDEC = (SYSDATE-2)
+          WHERE BFKEY = #{case_id}
+        SQL
+      end
+    end
+  end
+
+  def self.delete_test_data
+    # Only prepare test if there are less than 20 EstablishClaim tasks, as additional safeguard
+    # fail "Too many ClaimsEstablishment tasks" if EstablishClaim.count > 20
```
Comment (shanear) [inline app/services/test_data_service.rb]: here is a potential safeguard 🤔

File: app/views/test/setup/index.html.erb [hunk 7]
```diff
@@ -1,14 +1,39 @@
 <div class="cf-app-segment cf-app-segment--alt">
   <div class="usa-grid-full">
+    <% flash.each do |key,value| %>
```
Comment (shanear) [inline app/views/test/setup/index.html.erb]: Love this page ❤️

File: app/services/test_data_service.rb [hunk 8]
```diff
@@ -38,10 +39,10 @@ def self.prepare_claims_establishment!(vacols_id:, cancel_eps: false, decision_t
     vacols_case = VACOLS::Case.find(vacols_id)
     if decision_type == :full
       vacols_case.update_attributes(bfddec: AppealRepository.dateshift_to_utc(2.days.ago))
```
Comment (kierachell) [inline app/services/test_data_service.rb]: For consistency:
I was told by @joofsh that Caseworkers like to see _Decision date_ but we only deal with _Outcoding Date_, and Caseworkers would check them against each other - so if they see a discrepancy they'd freak out.

File: app/services/test_data_service.rb (hunk 6 above)
Comment (kierachell) [inline app/services/test_data_service.rb]: Seemed redundant (considering other safeguards) but I kept it in, just in case. Also, expecting to be able to process more than 20 decisions for testing (stress tests?)
I'll put it back in for now.

File: app/views/layouts/application.html.erb [hunk 9]
```diff
@@ -128,7 +128,7 @@
       <div>
         <div class="cf-push-left">
           <span<% if Rails.application.config.build_version %> title="<%= Rails.application.config.build_version[:date] %>"<% end %>>Built</span> with <abbr title="love">&#9825;</abbr> by the
-          <a href="http://www.va.gov/ds/">Digital Service at the <abbr title="Department of Veterans Affairs">VA</abbr></a>
+          <a href="https://www.va.gov/digitalstrategy/">Digital Service at the <abbr title="Department of Veterans Affairs">VA</abbr></a>
```
Comment (kierachell) [inline app/views/layouts/application.html.erb]: Thanks to @askldjd

File: app/controllers/test/users_controller.rb (hunk 2 above)
Comment (kierachell) [inline app/controllers/test/users_controller.rb]: @shanear 
We have two endpoints:
`test/users` for `dev`
`test/setup` for `uat`
I separated the functionality, previously - `test/users` would point to methods in `setup_controller`.
Next project: clean up `users_controller` by either combining it with `setup` or revamping it separately.

File: app/controllers/test/setup_controller.rb (hunk 4 above)
Comment (kierachell) [inline app/controllers/test/setup_controller.rb]: Just consistent name with our `PART_REMAND_IDS` environmental variable that limits which `VACOLS_IDS` we can use.
It's yet another safeguard - even if someone _could_ access this, they _cannot_ do anything outside of messing with our data.

File: app/services/test_data_service.rb (hunk 6 above)
Comment (shanear) [inline app/services/test_data_service.rb]: we can increase the # to 50 or 100, that will still miss prod.

File: app/controllers/test/setup_controller.rb (hunk 3 above)
Comment (kierachell) [inline app/controllers/test/setup_controller.rb]: For clarity: this whole page requires `uat` **and** `test_user`; but the main focus is that it is only available via `uat`.
The route is not even available if `env["TEST_USER_ID"]` is not set. So this is just a redundant check to make sure that even if `test_user_id` is set, then you still need to be in`uat` to do anything here.
Redundant checks are redundant, yes.

File: app/controllers/test/setup_controller.rb (hunk 4 above)
Comment (shanear) [inline app/controllers/test/setup_controller.rb]: right, but it's a confusing name. should be `PARTIAL_GRANT_AND_REMAND_IDS` or `PARTIAL_AND_REMAND_IDS`

File: app/controllers/test/users_controller.rb (hunk 2 above)
Comment (shanear) [inline app/controllers/test/users_controller.rb]: sounds good

File: app/services/test_data_service.rb (hunk 8 above)
Comment (shanear) [inline app/services/test_data_service.rb]: right, but you seem to be already updating the decision date here:
https://github.com/department-of-veterans-affairs/caseflow/pull/1163/files#diff-7d721dabb22976a610096677eae79e44R79

File: app/controllers/test/setup_controller.rb (hunk 5 above)
Comment (kierachell) [inline app/controllers/test/setup_controller.rb]: It is not available if `ENV["TEST_USER_ID"]` is not set on the server, and there's a `before_action` that checks if we're in `uat`.
I can change it so that it's a form that allows you to delete data for _one_ appeal at a time, would that work?
I basically do these actions via `uat` console anyway and it only deletes our data (doesn't touch VACOLS/VBMS or anything else).

File: app/controllers/test/setup_controller.rb (hunk 4 above)
Comment (kierachell) [inline app/controllers/test/setup_controller.rb]: Ok, I will update the ENV variable name as well.

File: app/services/test_data_service.rb (hunk 6 above)
Comment (kierachell) [inline app/services/test_data_service.rb]: 👍

File: app/services/test_data_service.rb (hunk 8 above)
Comment (kierachell) [inline app/services/test_data_service.rb]: yes I am 🤕 

will change

File: app/services/test_data_service.rb (hunk 6 above)
Comment (kierachell) [inline app/services/test_data_service.rb]: 👍 
also added another `test_user?` check just for fun

File: app/services/test_data_service.rb [hunk 10]
```diff
@@ -64,6 +65,29 @@ def self.cancel_end_products(appeal)
...
+          SET TIOCTIME = (SYSDATE-2)
+          WHERE TICKNUM = #{case_id}
+        SQL
+      end
+    end
+  end
+
+  def self.delete_test_data
+    # Only prepare test if there are less than 20 EstablishClaim tasks, as additional safeguard
+    fail "Too many ClaimsEstablishment tasks" if EstablishClaim.count > 50
```
Comment (kierachell) [inline app/services/test_data_service.rb]: restored and set to 50 to make sure it's available for future stress test; but won't affect prod

File: app/controllers/test/setup_controller.rb (hunk 5 above)
Comment (kierachell) [inline app/controllers/test/setup_controller.rb]: Safeguard restored;

**Comment by kierachell** [comment]
@shanear 
changes implemented
- safeguard restored for `too many claims`
- removed redundant decision date update
- clarified variable name
 let me know if I there's anything else I need to clean up;

File: app/models/appeal.rb (hunk 1 above)
Comment (joofsh) [inline app/models/appeal.rb]: ❤️ this change. Thanks artem!
